File: app/src/main/java/com/team2052/frckrawler/metric/view/MetricWidget.java
package com.team2052.frckrawler.metric.view;

import android.content.Context;
import android.view.LayoutInflater;
import android.widget.FrameLayout;

import com.google.gson.JsonElement;
import com.team2052.frckrawler.metric.data.MetricValue;
import com.team2052.frckrawler.models.Metric;

public abstract class MetricWidget extends FrameLayout {

    protected LayoutInflater inflater;
    private Metric metric;

    protected MetricWidget(Context context, MetricValue m) {
        this(context);
        metric = m.metric();
        setMetricValue(m);
    }

    protected MetricWidget(Context context) {
        super(context);
        inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        initViews();
    }

    public abstract void setMetricValue(MetricValue m);

    public abstract void initViews();

    public final MetricValue getValue() {
        return MetricValue.create(metric, getData());
    }

    public Metric getMetric() {
        return metric;
    }

    public abstract JsonElement getData();
}

File: core/metrics/src/main/java/com/team2052/frckrawler/core/metrics/view/MetricWidget.kt
package com.team2052.frckrawler.core.metrics.view

import android.content.Context
import android.view.LayoutInflater
import android.widget.FrameLayout
import com.google.gson.JsonElement
import com.team2052.frckrawler.core.data.models.Metric
import com.team2052.frckrawler.core.metrics.data.MetricValue

abstract class MetricWidget protected constructor(context: Context) : FrameLayout(context) {

    protected var inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    lateinit var metric: Metric

    protected constructor(context: Context, m: MetricValue) : this(context) {
        metric = m.metric
        setMetricValue(m)
    }

    init {
        initViews()
    }

    abstract fun setMetricValue(m: MetricValue)
    abstract fun initViews()

    fun getValue(): MetricValue = MetricValue(metric, data)

    abstract val data: JsonElement
}
